File: c_wrapper/pure_parser_examples.c
//
//  pure_parser_examples.c
//  PureParser
//
//  Created by Stan Potemkin on 05.12.2019.
//

#include "pure_parser.h"
#include <stdio.h>
#include <string.h>

#pragma mark - Local Types

typedef struct {
    const char *formula;
    char *output;
    const char *reference;
} test_result_t;

typedef void(*test_func_t)(test_result_t *);

typedef struct {
    const char *caption;
    test_func_t executor;
} test_meta_t;

#pragma mark - Local Helpers

static test_meta_t test_meta_make(const char *caption, test_func_t executor) {
    test_meta_t meta;
    meta.caption = caption;
    meta.executor = executor;
    return meta;
}

#pragma mark - Test Cases

static void test_Plain(test_result_t *result) {
    pure_config_t config;
    pure_config_set_default(&config);
    
    pure_parser_t parser;
    pure_parser_init(&parser, &config);
    
    const char *formula = "Hello world";
    result->formula = formula;
    result->reference = "Hello world";

    pure_parser_execute(&parser, result->formula, true, true, &result->output, NULL);
}

static void test_Variable(test_result_t *result) {
    pure_config_t config;
    pure_config_set_default(&config);
    
    pure_parser_t parser;
    pure_parser_init(&parser, &config);
    
    const char *formula = "Hello, $name";
    result->formula = formula;
    result->reference = "Hello, Stan";

    pure_parser_assign_var(&parser, "name", "Stan");
    pure_parser_execute(&parser, formula, true, true, &result->output, NULL);
}

static void test_VariableAndBlockWithVariable(test_result_t *result) {
    pure_config_t config;
    pure_config_set_default(&config);
    
    pure_parser_t parser;
    pure_parser_init(&parser, &config);
    
    const char *formula = "Please wait, $name: we're calling $[$anotherName ## another guy]";
    result->formula = formula;
    result->reference = "Please wait, Stan: we're calling Paul";

    pure_parser_assign_var(&parser, "name", "Stan");
    pure_parser_assign_var(&parser, "anotherName", "Paul");
    pure_parser_execute(&parser, result->formula, true, true, &result->output, NULL);
}

static void test_VariableAndBlockWithPlain(test_result_t *result) {
    pure_config_t config;
    pure_config_set_default(&config);
    
    pure_parser_t parser;
    pure_parser_init(&parser, &config);
    
    const char *formula = "Please wait, $name: we're calling $[$anotherName ## another guy]";
    result->formula = formula;
    result->reference = "Please wait, Stan: we're calling another guy";

    pure_parser_assign_var(&parser, "name", "Stan");
    pure_parser_execute(&parser, formula, true, true, &result->output, NULL);
}

static void test_BlockWithRich(test_result_t *result) {
    pure_config_t config;
    pure_config_set_default(&config);
    
    pure_parser_t parser;
    pure_parser_init(&parser, &config);
    
    const char *formula = "Congrats! You saved it $[in folder '$folder'].";
    result->formula = formula;
    result->reference = "Congrats! You saved it in folder 'Documents'.";

    pure_parser_assign_var(&parser, "folder", "Documents");
    pure_parser_execute(&parser, formula, true, true, &result->output, NULL);
}

static void test_InactiveBlock(test_result_t *result) {
    pure_config_t config;
    pure_config_set_default(&config);
    
    pure_parser_t parser;
    pure_parser_init(&parser, &config);
    
    const char *formula = "Congrats! You saved it $[in folder '$folder'].";
    result->formula = formula;
    result->reference = "Congrats! You saved it.";

    pure_parser_execute(&parser, result->formula, true, true, &result->output, NULL);
}

static void test_ComplexActiveAlias(test_result_t *result) {
    pure_config_t config;
    pure_config_set_default(&config);
    
    pure_parser_t parser;
    pure_parser_init(&parser, &config);
    
    const char *formula = "$[Agent $creatorName ## You] changed reminder $[«$comment»] $[:target: for $[$targetName ## you]] on $date at $time";
    result->formula = formula;
    result->reference = "You changed reminder «Check his payment» for you on today at 11:30 AM";

    pure_parser_assign_var(&parser, "comment", "Check his payment");
    pure_parser_assign_var(&parser, "date", "today");
    pure_parser_assign_var(&parser, "time", "11:30 AM");
    pure_parser_enable_alias(&parser, "target");
    pure_parser_execute(&parser, result->formula, true, true, &result->output, NULL);
}

static void test_ComplexInactiveAlias(test_result_t *result) {
    pure_config_t config;
    pure_config_set_default(&config);
    
    pure_parser_t parser;
    pure_parser_init(&parser, &config);
    
    const char *formula = "$[Agent $creatorName ## You] changed reminder $[«$comment»] $[:target: for $[$targetName ## you]] on $date at $time";
    result->formula = formula;
    result->reference = "You changed reminder «Check his payment» on today at 11:30 AM";

    pure_parser_assign_var(&parser, "comment", "Check his payment");
    pure_parser_assign_var(&parser, "date", "today");
    pure_parser_assign_var(&parser, "time", "11:30 AM");
    pure_parser_execute(&parser, result->formula, true, true, &result->output, NULL);
}

#pragma mark - Execute all Test Cases

#ifndef main_c
#define main_c main
#endif

int main_c() {
    #define declare_test_case(func) test_meta_make(#func, func)
    const test_meta_t all_tests[] = {
        declare_test_case(test_Plain),
        declare_test_case(test_Variable),
        declare_test_case(test_VariableAndBlockWithVariable),
        declare_test_case(test_VariableAndBlockWithPlain),
        declare_test_case(test_BlockWithRich),
        declare_test_case(test_InactiveBlock),
        declare_test_case(test_ComplexActiveAlias),
        declare_test_case(test_ComplexInactiveAlias)
    };
    #undef declare_test_case
    
    const size_t number_of_tests = sizeof(all_tests) / sizeof(*all_tests);
    test_result_t result;
    
    for (const test_meta_t *test = all_tests, *end = &all_tests[number_of_tests]; test < end; test++) {
        test->executor(&result);
        
        printf("Testing '%s'\n", test->caption);
        printf("> Formula: %s\n", result.formula);

        if (strcmp(result.output, result.reference) == 0) {
            printf("[OK]\n");
        }
        else {
            printf("> Output: %s\n", result.output);
            printf("> Reference: %s\n", result.reference);
            printf("[Failed]\n");
            return 1;
        }
    }

    printf("== All test cases passed OK ==\n");
    return 0;
}

File: swift_wrapper/PureParserExamples.swift
fileprivate struct Output {
    let formula: String
    let result: String
    let reference: String
    
    var hasPassed: Bool {
        return (result == reference)
    }
}

fileprivate func execute_Plain() -> Output {
    let parser = PureParser()
    let formula = "Hello world"
    
    return Output(
        formula: formula,
        result: parser.execute(formula, collapseSpaces: true, resetOnFinish: true),
        reference: "Hello world"
    )
}

fileprivate func execute_Variable() -> Output {
    let parser = PureParser()
    let formula = "Hello, $name"
    
    parser.assign(variable: "name", value: "Stan")
    
    return Output(
        formula: formula,
        result: parser.execute(formula, collapseSpaces: true, resetOnFinish: true),
        reference: "Hello, Stan"
    )
}

fileprivate func execute_VariableAndBlockWithVariable() -> Output {
    let parser = PureParser()
    let formula = "Please wait, $name: we're calling $[$anotherName ## another guy]"
    
    parser.assign(variable: "name", value: "Stan")
    parser.assign(variable: "anotherName", value: "Paul")

    return Output(
        formula: formula,
        result: parser.execute(formula, collapseSpaces: true, resetOnFinish: true),
        reference: "Please wait, Stan: we're calling Paul"
    )
}

fileprivate func execute_VariableAndBlockWithPlain() -> Output {
    let parser = PureParser()
    let formula = "Please wait, $name: we're calling $[$anotherName ## another guy]"
    
    parser.assign(variable: "name", value: "Stan")

    return Output(
        formula: formula,
        result: parser.execute(formula, collapseSpaces: true, resetOnFinish: true),
        reference: "Please wait, Stan: we're calling another guy"
    )
}

fileprivate func execute_BlockWithRich() -> Output {
    let parser = PureParser()
    let formula = "Congrats! You saved it $[in folder '$folder']."
    
    parser.assign(variable: "folder", value: "Documents")

    return Output(
        formula: formula,
        result: parser.execute(formula, collapseSpaces: true, resetOnFinish: true),
        reference: "Congrats! You saved it in folder 'Documents'."
    )
}

fileprivate func execute_InactiveBlock() -> Output {
    let parser = PureParser()
    let formula = "Congrats! You saved it $[in folder '$folder']."
    
    return Output(
        formula: formula,
        result: parser.execute(formula, collapseSpaces: true, resetOnFinish: true),
        reference: "Congrats! You saved it."
    )
}

fileprivate func execute_Complex() -> Output {
    let parser = PureParser()
    let formula = "$[Agent $creatorName ## You] changed reminder $[«$comment»] $[:target: for $[$targetName ## you]] on $date at $time"
    
    parser.assign(variable: "comment", value: "Check his payment")
    parser.assign(variable: "date", value: "today")
    parser.assign(variable: "time", value: "11:30 AM")
    parser.activate(alias: "target", true)

    return Output(
        formula: formula,
        result: parser.execute(formula, collapseSpaces: true, resetOnFinish: true),
        reference: "You changed reminder «Check his payment» for you on today at 11:30 AM"
    )
}

#if OUTER_EXECUTION
import PureParser

final class PureParserExamples {
    func run() {
        let cases = [
            ("test_Plain", execute_Plain),
            ("test_Variable", execute_Variable),
            ("test_VariableAndBlockWithVariable", execute_VariableAndBlockWithVariable),
            ("test_VariableAndBlockWithPlain", execute_VariableAndBlockWithPlain),
            ("test_BlockWithRich", execute_BlockWithRich),
            ("test_InactiveBlock", execute_InactiveBlock),
            ("test_Complex", execute_Complex)
        ]
        
        for (name, test) in cases {
            let out = test()
            print("Testing '\(name)'")
            print("> Formula: \(out.formula)")

            if out.hasPassed {
                print("[OK]")
            }
            else {
                print("> Output: \(out.result)")
                print("> Reference: \(out.reference)")
                print("[Failed]")
                assertionFailure()
            }

            print(" ")
        }
        
        print("== All test cases passed OK ==")
    }
}
#else
import XCTest

final class PureParserExamples: XCTestCase {
    static var allTests = [
        ("test_Plain", test_Plain),
        ("test_Variable", test_Variable),
        ("test_VariableAndBlockWithVariable", test_VariableAndBlockWithVariable),
        ("test_VariableAndBlockWithPlain", test_VariableAndBlockWithPlain),
        ("test_BlockWithRich", test_BlockWithRich),
        ("test_InactiveBlock", test_InactiveBlock),
        ("test_Complex", test_Complex),
    ]
    
    func test_Plain() {
        let output = execute_Plain()
        XCTAssertEqual(output.result, output.reference)
    }

    func test_Variable() {
        let output = execute_Variable()
        XCTAssertEqual(output.result, output.reference)
    }

    func test_VariableAndBlockWithVariable() {
        let output = execute_VariableAndBlockWithVariable()
        XCTAssertEqual(output.result, output.reference)
    }

    func test_VariableAndBlockWithPlain() {
        let output = execute_VariableAndBlockWithPlain()
        XCTAssertEqual(output.result, output.reference)
    }

    func test_BlockWithRich() {
        let output = execute_BlockWithRich()
        XCTAssertEqual(output.result, output.reference)
    }

    func test_InactiveBlock() {
        let output = execute_InactiveBlock()
        XCTAssertEqual(output.result, output.reference)
    }

    func test_Complex() {
        let output = execute_Complex()
        XCTAssertEqual(output.result, output.reference)
    }
}
#endif
